# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


In [20]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from statsmodels.stats import weightstats as stests
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

In [2]:
# Дані з умови
minutes = np.array([128, 135, 121, 142, 126, 151, 123])
mu0 = 123.7   # середнє у магазині

**1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.**

Нульова гіпотеза: середній час = 123,7, $\mu_0 = 123.7$

Альтернативна гіпотеза: середній час збільшився $\mu > 123.7$

Це правосторонній тест


2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.

In [10]:
s_mean = np.mean(minutes)
print(f"s_mean: {s_mean:.3f}")
s = np.std(minutes, ddof=1)
print(f"s: {s:.3f}")
n = np.size(minutes)
print(f"n: {n}")

s_mean: 132.286
s: 10.981
n: 7


3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.

In [ ]:
# Критичне значення для правостороннього тесту
df = n - 1
alpha = 0.1
t_crit = stats.t.ppf(1-alpha, df)
print(f"Критичне значення (df={df}): {t_crit:.3f},")

# scipy тест
res = stats.ttest_1samp(minutes, popmean=mu0, alternative="greater")
print(f"Статистика тесту = {res.statistic:.3f},\np-value = {res.pvalue:.3f}")

Критичне значення (df=6): 1.440,
Статистика тесту = 2.069,
p-value = 0.042


**4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.**

**5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?**

In [19]:
# висновок на основі р-значення
if res.pvalue < alpha:
    print(f"Відхиляємо H0: є статистично значущі докази, що після редизайну середній час до покупки збільшився ")
else:
    print("Не відхиляємо H0: доказів недостатньо")

Відхиляємо H0: є статистично значущі докази, що після редизайну середній час до покупки збільшився 


**6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.**

Рішення не змінюється - відхиляємо нульову гіпотезу.

Якби тест був двосторонній, то при $\alpha = 0.05$ не відхиляли б.

**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [25]:
x = 311
n = 1200
p0 = 0.23
alpha = 0.03

# z-test (правосторонній: p > p0)
z_stat, p_value = proportions_ztest(count=x, nobs=n, value=p0, alternative='larger')
confint = proportion_confint(count=x, nobs=n, alpha=alpha, method='normal')

print("Тест для однієї пропорції")
print(f"Z-статистика = {z_stat:.3f}")
print(f"p-value = {p_value:.3f}")
print(f"97% довірчий інтервал: {confint}")

Тест для однієї пропорції
Z-статистика = 2.306
p-value = 0.011
97% довірчий інтервал: (0.23171700302179205, 0.28661633031154127)


**Висновок:** є статистично значущі підстави стверджувати, що впізнаваність стала > 23% (тобто зросла).

Ми майже впевнені (на 97%), що реальна впізнаваність після кампанії десь між 23.17% і 28.66% - це вище за історичні 23%, тому кампанія (ймовірно) підвищила впізнаваність.